Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs

```diff
@@ -1,20 +1,239 @@
+            }
+
+            var containerName = CompOrNull<MetaDataComponent>(container)?.EntityName;
+            var dispenserName = CompOrNull<MetaDataComponent>(chemMaster.Owner)?.EntityName ?? "";
+            var bufferReagents = bufferSolution.Contents;
+            var bufferCurrentVolume = bufferSolution.CurrentVolume;
+
+            var state = new ChemMasterBoundUserInterfaceState(
+                containerSolution?.CurrentVolume, containerSolution?.MaxVolume, containerName, dispenserName,
+                containerSolution?.Contents, bufferReagents, chemMaster.Mode, bufferCurrentVolume, chemMaster.PillType,
+                chemMaster.PillProductionLimit, chemMaster.BottleProductionLimit, updateLabel
+            );
+            _userInterfaceSystem.TrySetUiState(chemMaster.Owner, ChemMasterUiKey.Key, state);
+        }
+
+        private void OnSetModeMessage(EntityUid uid, ChemMasterComponent chemMaster, ChemMasterSetModeMessage message)
+        {
+            // Ensure the mode is valid, either Transfer or Discard.
+            if (!Enum.IsDefined(typeof(ChemMasterMode), message.ChemMasterMode))
+                return;
+
+            chemMaster.Mode = message.ChemMasterMode;
+            UpdateUiState(chemMaster);
+            ClickSound(chemMaster);
+        }
+
+        private void OnSetPillTypeMessage(EntityUid uid, ChemMasterComponent chemMaster, ChemMasterSetPillTypeMessage message)
+        {
+            // Ensure valid pill type. There are 20 pills selectable, 0-19.
+            if (message.PillType > 19)
+                return;
+
+            chemMaster.PillType = message.PillType;
+            UpdateUiState(chemMaster);
+            ClickSound(chemMaster);
+        }
+
+        private void OnReagentButtonMessage(EntityUid uid, ChemMasterComponent chemMaster, ChemMasterReagentAmountButtonMessage message)
+        {
+            // Ensure the amount corresponds to one of the reagent amount buttons.
+            if (!Enum.IsDefined(typeof(ChemMasterReagentAmount), message.Amount))
+                return;
+
+            switch (chemMaster.Mode)
+            {
+                case ChemMasterMode.Transfer:
+                    TransferReagents(chemMaster, message.ReagentId, message.Amount.GetFixedPoint(), message.FromBuffer);
+                    break;
+                case ChemMasterMode.Discard:
+                    DiscardReagents(chemMaster, message.ReagentId, message.Amount.GetFixedPoint(), message.FromBuffer);
+                    break;
+                default:
+                    // Invalid mode.
+                    return;
+            }
+
+            ClickSound(chemMaster);
+        }
+
+        private void TransferReagents(ChemMasterComponent chemMaster, string reagentId, FixedPoint2 amount, bool fromBuffer)
+        {
+            var containerSlot = CompOrNull<ItemSlotsComponent>(chemMaster.Owner)?.Slots.GetValueOrDefault("container_slot");
```

Comment:
itemslotssystem like before. also do not hardcode this string value, make it a static/const somewhere.